### PR TITLE
TopDocs: ensure stable sorting on equal score

### PR DESCRIPTION
**Pull request by brainlock**
See gh-671

### CODE REVIEW

**Comment by brainlock** [comment]
If you look at the second commit, if we add the constraint that D is also PartialOrd in ComparableDoc, we can move the second ordering comparison directly in its implementation. Let me know whether you prefer solution 1 or 2, and I'll rebase.

File: src/collector/top_collector.rs [hunk 1]
```diff
@@ -175,6 +180,26 @@ impl<T: PartialOrd + Clone> TopSegmentCollector<T> {
     }
 }
 
+fn stable_sort<T, D>(a: &ComparableDoc<T, D>, b: &ComparableDoc<T, D>) -> Ordering
```
Comment (fulmicoton) [inline src/collector/top_collector.rs]: it's not a `stable_sort` so it should not be called stable sort.

Comment (fulmicoton) [inline src/collector/top_collector.rs]: I usually name that kind of argument `left` and `right`. A lot of ppl use `lhs`, and `rhs` (as left hand side, and right hand side). Let's avoid `a` and `b` anyway.

File: src/collector/top_collector.rs [hunk 2]
```diff
@@ -25,29 +24,41 @@ struct ComparableDoc<T, D> {
...
+impl<T: PartialOrd, D: PartialOrd> Ord for ComparableDoc<T, D> {
     #[inline]
     fn cmp(&self, other: &Self) -> Ordering {
-        other
+        // Reversed to make BinaryHeap work as a min-heap
+        let by_feature = other
             .feature
             .partial_cmp(&self.feature)
-            .unwrap_or_else(|| Ordering::Equal)
+            .unwrap_or_else(|| Ordering::Equal);
```
Comment (fulmicoton) [inline src/collector/top_collector.rs]: `unwrap_or` is better here.

Comment (fulmicoton) [reply src/collector/top_collector.rs]: Alternatively you could write it :

```rust

....
   .unwrap_or_else(lazy_by_doc_address)

```

File: src/collector/top_collector.rs [hunk 3]
```diff
@@ -175,6 +180,26 @@ impl<T: PartialOrd + Clone> TopSegmentCollector<T> {
...
+    D: PartialOrd,
+{
+    // Reversed since our ComparableDoc reverses the ordering (to
+    // make BinaryHeap work as a min-heap).
+    let by_feature = b
+        .feature
+        .partial_cmp(&a.feature)
+        .unwrap_or_else(|| Ordering::Equal);
+
+    let lazy_by_doc_address = || a.doc.partial_cmp(&b.doc).unwrap_or_else(|| Ordering::Equal);
```
Comment (fulmicoton) [inline src/collector/top_collector.rs]: `unwrap_or(Ordering::Equal)`..

The closure version is only useful if you do computation in it.

File: src/collector/top_collector.rs [hunk 4]
```diff
@@ -175,6 +180,26 @@ impl<T: PartialOrd + Clone> TopSegmentCollector<T> {
...
+where
+    T: PartialOrd,
+    D: PartialOrd,
+{
+    // Reversed since our ComparableDoc reverses the ordering (to
+    // make BinaryHeap work as a min-heap).
+    let by_feature = b
+        .feature
+        .partial_cmp(&a.feature)
+        .unwrap_or_else(|| Ordering::Equal);
```
Comment (fulmicoton) [inline src/collector/top_collector.rs]: no need for `_or_else`. See below.

**Comment by fulmicoton** [comment]
I added a couple of comments. 

The first version is less likely to induce a performance regression...
But the second version is actually easier to proofread and more robust to changes...

Let's check that there are no perf regression with version 2 and go with version 2. Do you know how to run the search benchmark?

**Comment by fulmicoton** [comment]
Can you add your change in the changelog ?
And don't forget your username... Boasting is our only salary in opensource ;)

**Comment by fulmicoton** [comment]
I don't think there is a `cargo bench` for the TopCollector. (I think there used to be one but I removed it).

I'm ok with `cargo bench` if you want such a bench. (that would be very useful for the future)
Otherwise, the search-benchmark-game.

**Comment by brainlock** [comment]
I ended up doing both, the search-benchmark-game felt like a useful "integration test".

The search-benchmark-game doesn't show any significant regression:
[search-benchmark-game-pr-675-against-master.tar.gz](https://github.com/tantivy-search/tantivy/files/3772334/search-benchmark-game-pr-675-against-master.tar.gz)
The reference run is the merge base of this PR.

Running the new benchmarks doesn't show any regression. Of the four, the one that does take longer in the PR (v2, so the latest commit) is the one for the cases where there are many ties on the score (`bench_top_segment_collector_collect_and_harvest_many_ties`), but that was expected as in that case more comparisons have to be done.

File: src/collector/top_collector.rs (hunk 1 above)
Comment (brainlock) [reply src/collector/top_collector.rs]: (this code doesn't exist anymore if we go for V2)

Comment (brainlock) [reply src/collector/top_collector.rs]: (this code doesn't exist anymore if we go for V2)

File: src/collector/top_collector.rs (hunk 2 above)
Comment (brainlock) [reply src/collector/top_collector.rs]: fixed

File: src/collector/top_collector.rs (hunk 3 above)
Comment (brainlock) [reply src/collector/top_collector.rs]: fixed

File: src/collector/top_collector.rs (hunk 4 above)
Comment (brainlock) [reply src/collector/top_collector.rs]: fixed

**Comment by brainlock** [comment]
@fulmicoton If you think the current diff is good to go, I'll rebase this into one commit.

**Comment by fulmicoton** [comment]
@brainlock Awesome thank you !

**Comment by brainlock** [comment]
Oops I wanted to rebase the ugly tmps and fixups away :/ Oh well too late :-)

Thanks for merging the pr!

**Pull request merged**